### PR TITLE
Add functionality to pin and unpin an idea.

### PR DESCRIPTION
**Pull request by asvinb**
## Summary

Addresses issue #3519 

## Relevant technical choices
Instead of dispatching the `fetchGetSavedIdeas()` and `fetchGetNewIdeas()` actions to refresh the data, the reducerCallback was used to manage the data.

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/idea-hub/datastore/idea-state.js [hunk 1]
```diff
@@ -49,6 +49,35 @@ const fetchPostUpdateIdeaStateStore = createFetchStore( {
...
+			};
+		}
+
+		if ( idea.saved === false ) {
+			return {
+				...state,
+				savedIdeas: ( state.savedIdeas || [] ).filter( ( { name } ) => name !== idea.name ),
+			};
+		}
+	},
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/idea-state.js:80]: Let's also return the `state` as is at the very end of the `reducerCallback` function.

File: assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/Idea.js [hunk 2]
```diff
@@ -49,23 +50,33 @@ import UnpinIcon from '../../../../../../svg/idea-hub-unpin.svg';
...
-	}, [ name ] );
-
-	const handleUnpin = useCallback( () => {
-		// @TODO: Implement callback.
-		global.console.log( `Unpinned: ${ name }` );
-	}, [ name ] );
+	const refreshNewSavedIdeas = useCallback( async () => {
+		await API.invalidateCache( 'modules', 'idea-hub', 'new-ideas' );
+		await API.invalidateCache( 'modules', 'idea-hub', 'saved-ideas' );
+	}, [] );
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/Idea.js:64]: Let's better move it to the `controlCallback` function of the `fetchPostUpdateIdeaStateStore` fetch store. We need to do the following:

1. Update the `controlCallback` function to return a new promise.
1. In the promise callback:
   1. Call and wait for results of the `API.set(...)` function.
   1. Call invalidate cache functions
   1. Resolve the promise passing the results of the `API.set(...)` function call.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb. It looks good to me. Added two suggestions and then it should be good to go.

File: assets/js/modules/idea-hub/datastore/idea-state.js (hunk 1 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/idea-state.js:87]: 👍

File: assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/Idea.js (hunk 2 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/Idea.js]: Updated @eugene-manuilov Can you check?

Comment (eugene-manuilov) [reply assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/Idea.js]: Yes, looks good to me. Thanks!

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb. It looks good to me. Could you please fix js tests? Once all checks pass, I'll approve and merge this PR.

**Comment by asvinb** [comment]
@eugene-manuilov Failing test fixed. Can you check whether the approach is good?
Thanks!

File: assets/js/modules/idea-hub/datastore/idea-state.js [hunk 3]
```diff
@@ -32,23 +32,63 @@ const { receiveError, clearError } = errorStoreActions;
...
+
+		try {
+			const results = await API.set( 'modules', 'idea-hub', 'update-idea-state', params );
+			await API.invalidateCache( 'modules', 'idea-hub', 'new-ideas' );
+			await API.invalidateCache( 'modules', 'idea-hub', 'saved-ideas' );
+
+			resolve( results );
+		} catch ( err ) {
+			reject( err );
+		}
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/idea-state.js:52]: Actually, I think we can simplify it once again and get rid of `new Promise` approach. How about doing this?

```javascript
controlCallback: ( { name, saved, dismissed } ) => {
    const params = { name };

    if ( saved !== undefined ) {
        params.saved = saved;
    } else {
        params.dismissed = dismissed;
    }

    return API.set( 'modules', 'idea-hub', 'update-idea-state', params ).then( () => async {
        await API.invalidateCache( 'modules', 'idea-hub', 'new-ideas' ).catch( () => {} );
        await API.invalidateCache( 'modules', 'idea-hub', 'saved-ideas' ).catch( () => {} );
    } );
},
```

This should also work, shouldn't it? Sorry, for the incorrect suggestion at the beginning.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]

File: assets/js/modules/idea-hub/datastore/idea-state.js (hunk 3 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/idea-state.js]: Yep, works with a few tweaks from your suggestion :)

Comment (eugene-manuilov) [reply assets/js/modules/idea-hub/datastore/idea-state.js]: Ah... yes, you are right, we need to return the result :laughing:

File: assets/js/modules/idea-hub/datastore/idea-state.js [hunk 4]
```diff
@@ -32,15 +32,21 @@ const { receiveError, clearError } = errorStoreActions;
 
 const fetchPostUpdateIdeaStateStore = createFetchStore( {
 	baseName: 'updateIdeaState',
-	controlCallback: ( { name, saved, dismissed } ) => {
+	controlCallback: async ( { name, saved, dismissed } ) => {
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/idea-state.js:35]: This `async` isn't needed here:
```suggestion
	controlCallback: ( { name, saved, dismissed } ) => {
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb! Looks good to me. Left one suggestion that I'll apply by myself.

**Review by eugene-manuilov — APPROVED** [review]

**Pull request merged**